qiskit/circuit/random/utils.py

```diff
@@ -11,15 +11,394 @@
+                        qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                elif control_qubit not in qubit_idx_used and target_qubit in qubit_idx_used:
+                    if insert_1q_oper:
+                        qc._append(
+                            CircuitInstruction(
+                                operation=_get_extra_1q_oper(extra_1q_gates, parameters_1q),
+                                qubits=[qubits[control_qubit]],
+                            )
+                        )
+                        qubit_idx_used.update({control_qubit})
+                        qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                elif control_qubit in qubit_idx_used and target_qubit in qubit_idx_used:
+                    continue
+
+                else:
+                    qc._append(
+                        CircuitInstruction(
+                            operation=operation,
+                            qubits=[qubits[control_qubit], qubits[target_qubit]],
+                        )
+                    )
+                    qubit_idx_used.update(set(edge))
+                    qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                    edges_used[(control_qubit, target_qubit)] += 1
+
+        else:
+            for current_gate, num_gate_params, edge in zip(
+                gate_choices["class"],
+                gate_choices["num_params"],
+                edge_choices,
+            ):
+                params = parameters[:num_gate_params]
+                parameters = parameters[num_gate_params:]
+                operation = current_gate(*params)
+                control_qubit, target_qubit = tuple(edge)
+
+                if control_qubit in qubit_idx_used and target_qubit not in qubit_idx_used:
+                    if insert_1q_oper:
+                        qc._append(
+                            CircuitInstruction(
+                                operation=_get_extra_1q_oper(extra_1q_gates, parameters_1q),
+                                qubits=[qubits[target_qubit]],
+                            )
+                        )
+                        qubit_idx_used.update({target_qubit})
+                        qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                elif control_qubit not in qubit_idx_used and target_qubit in qubit_idx_used:
+                    if insert_1q_oper:
+                        qc._append(
+                            CircuitInstruction(
+                                operation=_get_extra_1q_oper(extra_1q_gates, parameters_1q),
+                                qubits=[qubits[control_qubit]],
+                            )
+                        )
+                        qubit_idx_used.update({control_qubit})
+                        qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                elif control_qubit in qubit_idx_used and target_qubit in qubit_idx_used:
+                    continue
+
```

Comment:
Why couldn't you reuse the slack logic in the original `random_circuit`? I'm not sure this code is correct - what about qubits that are neither on a target/control qubit, is this possible?

Reply:
Yes, indeed, the logic was wrong, I have changed the logic. Its explained in the docstring.